services/brig/index/src/Brig/Index/Migrations.hs

```diff
@@ -0,0 +1,141 @@
+import qualified Database.V5.Bloodhound as ES
+import Imports
+import qualified Network.HTTP.Client as HTTP
+import System.Logger.Class (Logger)
+import qualified System.Logger.Class as Log
+
+migrate :: Logger -> Opts.ElasticSettings -> Opts.CassandraSettings -> IO ()
+migrate l es cas = do
+  env <- mkEnv l es cas
+  finally (go env) (cleanup env)
+  where
+    go env =
+      runMigrationAction env $ do
+        failIfIndexAbsent (es ^. Opts.esIndex)
+        createMigrationsIndexIfNotPresent
+        runMigrations [V1_ReIndexForSearchTeamMembers.migration]
+
+indexName :: ES.IndexName
+indexName = ES.IndexName "wire_brig_migrations"
+
+indexMappingName :: ES.MappingName
+indexMappingName = ES.MappingName "wire_brig_migrations"
+
+indexMapping :: Value
+indexMapping =
+  object
+    [ "properties"
+        .= object
+          ["migration_version" .= object ["index" .= True, "type" .= ("integer" :: Text)]]
+    ]
+
+mkEnv :: Logger -> Opts.ElasticSettings -> Opts.CassandraSettings -> IO Env
+mkEnv l es cas =
+  Env
+    <$> initES
+      <*> initCassandra
+      <*> initLogger
+      <*> Metrics.metrics
+      <*> (pure $ view Opts.esIndex es)
+  where
+    initES =
+      ES.mkBHEnv
+        (Opts.toESServer (es ^. Opts.esServer))
+        <$> HTTP.newManager HTTP.defaultManagerSettings
+    initCassandra =
+      C.init
+        $ C.setLogger (C.mkLogger l)
+          . C.setContacts (view Opts.cHost cas) []
+          . C.setPortNumber (fromIntegral (view Opts.cPort cas))
+          . C.setKeyspace (view Opts.cKeyspace cas)
+          . C.setProtocolVersion C.V4
+        $ C.defSettings
+    initLogger = pure l
+
+createMigrationsIndexIfNotPresent :: (MonadThrow m, MonadIO m) => MigrationActionT m ()
+createMigrationsIndexIfNotPresent =
+  liftBH $ do
+    unlessM (ES.indexExists indexName) $
+      ES.createIndexWith [] 1 indexName
+        >>= throwIfNotCreated CreateMigrationIndexFailed
+    ES.putMapping indexName indexMappingName indexMapping
+      >>= throwIfNotCreated PutMappingFailed
+  where
+    throwIfNotCreated err response =
+      unless (ES.isSuccess response)
+        $ throwM
+        $ err (show response)
+
+failIfIndexAbsent :: (MonadThrow m, MonadIO m) => ES.IndexName -> MigrationActionT m ()
+failIfIndexAbsent targetIndex =
+  liftBH $
+    unlessM
+      (ES.indexExists targetIndex)
+      (throwM $ TargetIndexAbsent targetIndex)
+
+-- | Runs only the migrations which need to run
+runMigrations :: [Migration] -> MigrationActionT IO ()
+runMigrations migrations = do
+  vmax <- latestMigrationVersion
+  let pendingMigrations = filter (\m -> version m > vmax) migrations
+  if null pendingMigrations
+    then info "No new migrations."
+    else info "New migrations found."
+  mapM_ runMigration pendingMigrations
+
+runMigration :: Migration -> MigrationActionT IO ()
+runMigration (Migration ver txt mig) = do
+  info $ "Running: [" <> show (migrationVersion ver) <> "] " <> Text.unpack txt
+  mig
+  persistVersion ver
```

Comment:
Are migrations idempotent?

Reply:
The one that we have, is idempotent. Hopefully the future ones are too. If not we'd have to have a more sophisticated mechanism to make sure we never re-run a failed migration.